### PR TITLE
Enable eclipselink with H2 database in testing

### LINKED ISSUE #149: [FEATURE REQUEST] Switch the tests against EclipseLink with H2 database
### Is your feature request related to a problem? Please describe.

Currently, the tests are tested against InMemoryPolarisMetaStoreManagerFactory and PolarisTreeMapMetaStoreSessionImpl implementation only and only a unit test is covering for EclipseLinkPolarisMetaStoreManagerFactory. To be closer to the production with database backed EclipseLinkPolarisMetaStoreManagerFactory and PolarisEclipseLinkMetaStoreSessionImpl, we should also consider adding all the tests against EclipseLinkPolarisMetaStoreManagerFactory. 

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aihuaxu**
# Description

The change is to support the testing against eclipselink with file-based H2 database in addition to in-memory implementation. That will test the env closer to the production. 
There is a difference between in-memory metastore and eclipseLink metastore: in-memory gets bootstrapped automatically since it's only used for test purpose while eclipseLink metastore doesn't boostrap automatically to avoid accidental wipeout in production.

To test out against  eclipseLink metastore, 
- The metastore gets bootstrapped for non in-memory metastore in tests (PolarisConnectionExtension).  
- The test DB location is placed in testDir and they will be cleaned up when the test starts to make the tests clean. 
- Remove persistence.xml in test and only keep one persistence.xml in main. I don't see value to keep two since the production version will always be a external version. 

TODO:

- There are some tests currently only testing against in-memory. We will need to add the coverage for elicplseLink.
- Instead of adding H2 dependencies to polaris-service, we can support loading the additional dependencies ondemand.

Fixes #149 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] This has been tested against with the existing tests `./gradlew test` and all the tests passed against eclipselink.
- [x] This has been tested against with the existing tests `./gradlew test` and all the tests passed against in-memory.

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

Please delete options that are not relevant.

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue
- [ ] I have signed and submitted the [ICLA](https://github.com/polaris-catalog/polaris/blob/main/ICLA.md) and if needed, the [CCLA](https://github.com/polaris-catalog/polaris/blob/main/CCLA.md). See [Contributing](https://github.com/polaris-catalog/polaris/blob/main/CONTRIBUTING.md) for details.

### CODE REVIEW

File: polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java [hunk 1]
```diff
@@ -102,12 +110,18 @@ public void purgeRealms(List<String> realms) {
...
   }
 
   @Override
   public synchronized PolarisMetaStoreManager getOrCreateMetaStoreManager(
       RealmContext realmContext) {
+    String realmId = realmContext.getRealmIdentifier();
+    if (bootstrapEnabled && !bootstrappedRealms.contains(realmId)) {
+      bootstrapRealmAndPrintCredentials(realmId);
+    }
+
```
Comment (collado-mike) [inline polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:124]: This code was intentionally moved out of here. Nobody wants to accidentally bootstrap their production instance. Including this code here allows for a misconfigured server instance to irrevocably wipe out all production data. Instead, let's just update the tests to explicitly call bootstrap on the metastore manager

**Review by collado-mike — CHANGES_REQUESTED** [review]
I think the `PolarisConnectionExtension` class is the best place to enable bootstrapping the metastore for tests. When we added the `bootstrap` command, we explicitly removed this logic from the persistent metastore impls so that users couldn't accidentally bootstrap their production instance.

File: polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java (hunk 1 above)
Comment (sfc-gh-aixu) [reply polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:124]: That makes sense.

**Comment by RussellSpitzer** [comment]
@aihuaxu Could you please rebase this?

**Comment by aihuaxu** [comment]
> @aihuaxu Could you please rebase this?

@RussellSpitzer Rebased. Please help take a look.

File: extension/persistence/eclipselink/src/main/java/org/apache/polaris/extension/persistence/impl/eclipselink/PolarisEclipseLinkMetaStoreSessionImpl.java [hunk 2]
```diff
@@ -153,8 +154,8 @@ private EntityManagerFactory createEntityManagerFactory(
           prefixUrl = new File(jarPrefixPath).toURI().toURL();
         }
 
-        LOGGER.info(
-            "Created a new ClassLoader with the jar {} in classpath to load the config file",
+        LOGGER.debug(
+            "Create a new ClassLoader with the jar {} in classpath to load the config file",
```
Comment (eric-maynard) [inline extension/persistence/eclipselink/src/main/java/org/apache/polaris/extension/persistence/impl/eclipselink/PolarisEclipseLinkMetaStoreSessionImpl.java:158]: ultra nit: `Creating`?

File: polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java [hunk 3]
```diff
@@ -46,7 +46,6 @@
  */
 public abstract class LocalPolarisMetaStoreManagerFactory<StoreType>
     implements MetaStoreManagerFactory {
-
```
Comment (eric-maynard) [inline polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:49]: nit: spurious change?

File: polaris-service/build.gradle.kts [hunk 4]
```diff
@@ -108,6 +108,9 @@ dependencies {
   testImplementation(libs.assertj.core)
   testImplementation(libs.mockito.core)
   testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+
+  testRuntimeOnly(project(":polaris-eclipselink"))
```
Comment (eric-maynard) [inline polaris-service/build.gradle.kts:112]: For now I think this could be okay, but I feel like we may not want these dependencies here forever. We have intentionally not included these dependencies in the service. A new project to include all these dependencies could make sense.

File: polaris-service/src/test/java/org/apache/polaris/service/admin/PolarisOverlappingCatalogTest.java [hunk 5]
```diff
@@ -110,101 +119,72 @@ private Response createCatalog(
...
-                            .returns(
-                                Response.Status.BAD_REQUEST.getStatusCode(), Response::getStatus);
-
-                        // `root` is inside this
-                        assertThat(createCatalog(prefix, "", laterExternal))
-                            .returns(
-                                Response.Status.BAD_REQUEST.getStatusCode(), Response::getStatus);
-                      });
-            });
+  @ParameterizedTest
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/admin/PolarisOverlappingCatalogTest.java:122]: Are these changes related to the PR?

File: polaris-service/src/test/java/org/apache/polaris/service/test/PolarisConnectionExtension.java [hunk 6]
```diff
@@ -109,10 +118,40 @@ public void beforeAll(ExtensionContext extensionContext) throws Exception {
...
+    }
+  }
+
   public static String getTestRealm(Class testClassName) {
     return testClassName.getName().replace('.', '_');
   }
 
+  public static void createTestDir(String realm) throws IOException {
+    // Set up the database location
+    Path testDir = Path.of("build/test_data/iceberg/" + realm);
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/test/PolarisConnectionExtension.java:134]: iceberg?

File: polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java (hunk 3 above)
Comment (aihuaxu) [reply polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:49]: Yeah.  Let me remove this unrelated change from this PR.

File: polaris-service/build.gradle.kts (hunk 4 above)
Comment (aihuaxu) [reply polaris-service/build.gradle.kts:112]: If we create a separate project to include the dependencies, we still need to include such project here , right? This is for test only though.

File: polaris-service/src/test/java/org/apache/polaris/service/admin/PolarisOverlappingCatalogTest.java (hunk 5 above)
Comment (aihuaxu) [reply polaris-service/src/test/java/org/apache/polaris/service/admin/PolarisOverlappingCatalogTest.java:122]: It's not. Let me revert.

File: polaris-service/src/test/java/org/apache/polaris/service/test/PolarisConnectionExtension.java (hunk 6 above)
Comment (aihuaxu) [reply polaris-service/src/test/java/org/apache/polaris/service/test/PolarisConnectionExtension.java:134]: Yeah. That is like this currently. Let me change to polaris.

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java [hunk 7]
```diff
@@ -288,7 +289,7 @@ public void run(PolarisApplicationConfig configuration, Environment environment)
     // different processes
     // and in-memory state will be lost b/w invocation of bootstrap command and running a server
```
Comment (flyrain) [inline polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:290]: Nit: Can we format the comment a bit since we are here. Not a blocker though.

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java [hunk 8]
```diff
@@ -288,7 +289,7 @@ public void run(PolarisApplicationConfig configuration, Environment environment)
...
     if (metaStoreManagerFactory instanceof InMemoryPolarisMetaStoreManagerFactory) {
-      metaStoreManagerFactory.getOrCreateMetaStoreManager(configuration::getDefaultRealm);
+      metaStoreManagerFactory.bootstrapRealms(List.of(configuration.getDefaultRealm()));
     }
```
Comment (flyrain) [inline polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:293]: Can we move this logic into class `InMemoryPolarisMetaStoreManagerFactory` as a part of initialization? Actually it is already there, but deleted in this PR. IIUC, here is the logic. We can make this change due to `getOrCreateMetaStoreManager()` is invoked already before this line. In that case, why do we need to bootstrap again here? We can bootstrap in the method `InMemoryPolarisMetaStoreManagerFactory::getOrCreateMetaStoreManager`

File: polaris-service/src/test/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandlerTest.java [hunk 9]
```diff
@@ -43,15 +44,22 @@
 import org.apache.polaris.core.entity.AsyncTaskType;
 import org.apache.polaris.core.entity.TaskEntity;
 import org.apache.polaris.service.persistence.InMemoryPolarisMetaStoreManagerFactory;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 class ManifestFileCleanupTaskHandlerTest {
+  private InMemoryPolarisMetaStoreManagerFactory metaStoreManagerFactory;
+  RealmContext realmContext;
```
Comment (flyrain) [inline polaris-service/src/test/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandlerTest.java:52]: make it private ?

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java (hunk 8 above)
Comment (aihuaxu) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:293]: I want to make the behavior the same regardless of InMemory or EclipseLink so we can test against both with the same set of tests. We will not bootstrap automatically but bootstrap if needed - we bootstrap in the tests for both in-memory and EclipseLink and here in service against in-memory.

Comment (flyrain) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:293]: I understand that we want to explicitly bootstrap in test, which applies to both in-memory and eclipse-link. But this place is a bit different. I was wondering if there is a better way to handle bootstrapping here. Ideally, it should be an internal operation within `InMemoryPolarisMetaStoreManagerFactory`.

**Review by flyrain — COMMENTED** [review]
Thanks @aihuaxu for the PR. LGTM overall, left minor comments.

**Review by flyrain — APPROVED** [review]

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java (hunk 8 above)
Comment (aihuaxu) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java]: I reverted the change actually since the users actually rely on the output to get the credentials from console.

File: polaris-service/src/main/java/org/apache/polaris/service/persistence/InMemoryPolarisMetaStoreManagerFactory.java [hunk 10]
```diff
@@ -48,40 +40,4 @@ protected PolarisMetaStoreSession createMetaStoreSession(
...
-
-  @Override
-  public synchronized PolarisMetaStoreManager getOrCreateMetaStoreManager(
-      RealmContext realmContext) {
-    String realmId = realmContext.getRealmIdentifier();
-    if (!bootstrappedRealms.contains(realmId)) {
-      bootstrapRealmAndPrintCredentials(realmId);
-    }
-    return super.getOrCreateMetaStoreManager(realmContext);
-  }
```
[the remark this answers is not in the document]
Comment (aihuaxu) [reply polaris-service/src/main/java/org/apache/polaris/service/persistence/InMemoryPolarisMetaStoreManagerFactory.java]: I brought back the change now.

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java (hunk 7 above)
Comment (aihuaxu) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java]: I reverted the change so it's not updated anymore. I will update for now.

File: polaris-service/build.gradle.kts (hunk 4 above)
Comment (aihuaxu) [reply polaris-service/build.gradle.kts:112]: I will try to address in the following PRs to include the dependency if needed. I saw that we have `-PeclipseLink=$ECLIPSELINK`.

**Comment by sfc-gh-aixu** [comment]
@RussellSpitzer, @eric-maynard  and @collado-mike Can you help review again?

**Comment by flyrain** [comment]
retest this please

**Review by flyrain — APPROVED** [review]
+1

**Comment by flyrain** [comment]
Thanks @aihuaxu for working on it. Thanks @collado-mike @RussellSpitzer @eric-maynard for the review.

**Review by flyrain — APPROVED** [review]

**Comment by aihuaxu** [comment]
Thanks @flyrain.
